### PR TITLE
Fix many Dialyzer warnings

### PR DESCRIPTION
**Pull request by essen**
Fixed nearly 170 compared to when I started (though a couple fixes were done by @dumbbell as well and already in master). On the other hand after rebasing around 60-70 new warnings popped up. A good number of them are probably related to a handful of errors though so they don't look too hard to fix.

I left a few things commented out. Please express your opinion on whether that code is really dead and/or really should be dead and I'll remove/fix it as necessary.

## Proposed Changes

Please describe the big picture of your changes here to communicate to the
RabbitMQ team why we should accept this pull request. If it fixes a bug or
resolves a feature request, be sure to link to that issue.

A pull request that doesn't explain **why** the change was made has a much
lower chance of being accepted.

If English isn't your first language, don't worry about it and try to
communicate the problem you are trying to solve to the best of your abilities.
As long as we can understand the intent, it's all good.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [ ] I have read the `CONTRIBUTING.md` document
- [ ] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc.

### CODE REVIEW

File: src/rabbit_reader.erl [hunk 1]
```diff
@@ -191,7 +191,7 @@ init(Parent, HelperSup, Ref) ->
 system_continue(Parent, Deb, {Buf, BufLen, State}) ->
     mainloop(Deb, Buf, BufLen, State#v1{parent = Parent}).
 
--spec system_terminate(_,_,_,_) -> none().
+-spec system_terminate(_,_,_,_) -> no_return().
```
Comment (essen) [inline src/rabbit_reader.erl]: `none()` and `no_return()` should be equivalent. But somehow only `no_return()` fixes the Dialyzer warning. If anyone has an explication...

**Comment by michaelklishin** [comment]
@essen @dumbbell @kjnilsson I merged master into this branch and resolved conflicts to the best of my ability. Since specs were moved to definitions in several modules in `master` it was too time consuming to check every single function but we don't have to do it all at once in this PR. Please take a look.

**Comment by essen** [comment]
The above comment about the merge/spec position should not apply anymore.

**Comment by michaelklishin** [comment]
I deleted the now outdated comment, thank you, @essen .

**Comment by michaelklishin** [comment]
This passes all core and plugin test suites I tried (about 10 of them) 👍

**Pull request merged**